findbugs/src/java/edu/umd/cs/findbugs/detect/IdentityHashCodeOnValueBasedClass.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * FindBugs - Find Bugs in Java programs
+ * Copyright (C) 2003-2008 University of Maryland
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+package edu.umd.cs.findbugs.detect;
+
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.OpcodeStack.Item;
+import edu.umd.cs.findbugs.ba.vbc.ValueBasedClassIdentifier;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+import edu.umd.cs.findbugs.util.ClassName;
+
+public class IdentityHashCodeOnValueBasedClass extends OpcodeStackDetector {
+
+    private final BugReporter bugReporter;
+
+    public IdentityHashCodeOnValueBasedClass(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+    }
+
```

Comment:
You may quickly bypass uninteresting classes which will make this detector faster.

1. Create a constant method descriptor:

        private static final MethodDescriptor IDENTITY_HASH_CODE = 
                new MethodDescriptor("java/lang/System", "identityHashCode", "(Ljava/lang/Object;)I", true);

2. Add the following check:

        @Override
        public void visitClassContext(ClassContext classContext) {
            if(hasInterestingMethod(classContext.getJavaClass().getConstantPool(), 
                                                       Collections.singleton(IDENTITY_HASH_CODE))) {
                super.visitClassContext(classContext);
            }
        }

So if class constant pool has no identityHashCode method, you skip the whole class which would be much faster than examining its code.